rfcs/text/0003_application_service_mounting.md

```diff
@@ -0,0 +1,259 @@
+- Start Date: 2019-05-10
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A front-end service to manage registration and root-level routing for
+first-class applications.
+
+# Basic example
+
+```tsx
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { MyApp } from './components';
+
+class MyPlugin {
+  setup({ application }) {
+    application.registerApp({
+      id: 'my-app',
+      title: 'My Application',
+      rootRoute: '/myapp',
+      mount(targetDomElement, pluginStartContext, pluginStart) {
+        ReactDOM.render(
+          <MyAppRoot core={pluginStartContext} deps={pluginStart} />,
+          targetDomElement
+        );
+
+        return () => {
+          ReactDOM.unmountComponentAtNode(targetDomElement);
+        };
+      }
+    });
+  }
+}
+```
+
+# Motivation
+
+By having centralized management of applications we can have a true single page
+application. It also gives us a single place to enforce authorization and/or
+licensing constraints on application access.
+
+By making the mounting interface of the ApplicationService generic, we can
+support many different rendering technologies simultaneously to avoid framework
+lock-in.
+
+# Detailed design
+
+This is the bulk of the RFC. Explain the design in enough detail for somebody
+familiar with Kibana to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+## Interface
+
+```ts
+export type Unmount = () => void;
+
+export interface AppSpec {
+  id: string;
+
+  /**
+   * The title of the application.
+   */
+  title: string;
+
+  /**
+   * The root route to mount this application at. Must be unique across all
+   * registered applications.
+   */
+  rootRoute: string;
+
+  /**
+   * A mount function called when the user navigates to this app's `rootRoute`.
+   * @param targetDomElement An HTMLElement to mount the application onto.
+   * @returns An unmounting function that will be called to unmount the application.
+   */
+  mount(targetDomElement: HTMLElement): Unmount
+
+  /**
+   * A EUI iconType that will be used for the app's icon. This icon
+   * takes precendence over the `icon` property.
+   */
+  euiIconType?: string;
+
+  /**
+   * A URL to an image file used as an icon. Used as a fallback
+   * if `euiIconType` is not provided.
+   */
+  icon?: string;
```

Comment:
Would be nice if we had an access to the icon DOM element, so we can render anything in the icon.

Maybe `icon` could also accept a function, which would receive a DOM element as its single argument.

```ts
{
  icon: (div: HTMLDivElement) => {
    render(<SuperIcon />, div);
  }
}
```

That way apps could have really fancy icons. Like a progress bar could be in an icon:

![image](https://user-images.githubusercontent.com/9773803/58382157-193a9580-7fc7-11e9-9b64-10a0251c0507.png)

Or icon could have a badge in a corner:

![image](https://user-images.githubusercontent.com/9773803/58382163-38392780-7fc7-11e9-877e-7e391fe5485b.png)

![image](https://user-images.githubusercontent.com/9773803/58382170-5010ab80-7fc7-11e9-89fd-5bd4b42523f9.png)

Really anything could be added to icons as apps would be in control of their icon look.

Reply:
![image](https://user-images.githubusercontent.com/9773803/58384661-9fb1a000-7fe4-11e9-96bb-71afb2717f1f.png)

Reply:
@streamich Would you be interested in submitting a feature request for these ideas on the [EUI repo](http://github.com/elastic/eui)? If EUI published a `SideBarIcon` component then it could add props to support these and other features in the future. If each app rendered this component then it would have access to these new features for free as they get added.

Reply:
I think this is something we could add in the future in a non-breaking way but is not critical to support at this time.